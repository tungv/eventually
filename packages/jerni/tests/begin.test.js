const createJourney = require("../lib/createJourney2");
const makeServer = require("./makeServer");

test("#begin() should push events to stores", async () => {
  const { server } = await makeServer({
    ns: "test-begin-method",
    port: 19080,
  });

  try {
    const store = makeTestStore(event => event.id);

    const journey = createJourney({
      writeTo: "http://localhost:19080",
      stores: [store],
    });

    await journey.commit({ type: "type_1", payload: {} });
    await journey.commit({ type: "type_2", payload: {} });
    await journey.commit({ type: "type_3", payload: {} });

    const db = await store.getDriver();
    const outputs = [];
    for await (const output of journey.begin()) {
      // only 2 possible events because we have filters by type
      outputs.push(output);
      if (db.length === 2) {
        expect(db).toEqual([1, 2]);
        break;
      }
    }

    expect(outputs).toEqual([["done 2"]]);

    // journey.destroy();
  } finally {
    server.close();
  }
});

function makeTestStore(transform) {
  const db = [];
  let listeners = [];

  const store = {
    name: "test_store",
    meta: {
      includes: ["type_1", "type_2"],
    },
    registerModels(map) {},
    subscribe(listener) {
      listeners.push(listeners);

      return () => {
        listeners = listeners.filter(fn => fn !== listener);
      };
    },
    async handleEvents(events) {
      db.push(...events.map(transform));
      listeners.forEach(fn => fn(last(events).id));
      return `done ${last(events).id}`;
    },

    async getDriver() {
      return db;
    },

    async getLastSeenId() {
      const event = last(db);
      if (event) return event.id;
      return 0;
    },
  };

  return store;
}

const last = array => (array.length >= 1 ? array[array.length - 1] : null);
