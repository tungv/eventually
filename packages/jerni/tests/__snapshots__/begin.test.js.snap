// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#begin() should push events to stores 1`] = `
"[DBG] === SWITCH TO NEW LOGGER PROVIDED BY begin() ===
[DBG] journey.begin({{
  pulseCount: 200,
  pulseTime: 10,
  serverUrl: undefined,
  cleanStart: false
}})
[DBG] sending http request to: http://localhost:19080/subscribe
[DBG] headers {
  'last-event-id': '0',
  includes: 'type_1,type_2',
  'burst-count': 200,
  'burst-time': 10
}
[DBG] socket opened!
[INF] start receiving data
[DBG] handling events #1 - #2
[DBG] done
[INF] stop processing events"
`;
